### ISSUE TITLE
Development cartridge for Thomson MO5

### ISSUE THREAD

**Issue by sporniket** [sporniket/ideas#48]
# Overview

* A cartridge holding a 5V NOR flash (like Microchip SST39Fxxx series (SST39F010, SST39F020, and SST39F040 for 128kB, 256kB and 512kB) to be plugged into the cartridge port of a Thomson MO5 computer.
* holds Arduino nano with a firmware allowing to receive a firmware to write to the flash through the USB connection to a host computer using the serial link
* All circuitry required to : 
  * electrically isolate the cart from either the programming circuit ("EXEC mode") or the computer ("PROGRAMMING mode")
  * allow the arduino nano to program the flash (shift registers to save pin usage)
  * optionnally, make a part of the Flash writable while in "EXEC mode", to act as a saving slot, for games.
  * some LEDs to indicate current mode and state (EXEC/PROGRAMMING-{wait, erase, write, check, error}

# User stories

### Updating the content of the cartridge

* **I want to** flash an image file representing the content of the cartridge
  * from my computer, plugged to the cartridge using a standard USB cable (Host to device, mini B, micro B or micro C)
  * Using either a standard tool like [openocd](https://openocd.org/) or an adhoc program to write
  * with the cartridge plugged in the Thomson MO5, that computer being switched on
* **In order to** continuously write software on my computer, build, send to the cartridge, reset the Thomson and test

### Allow the software to save some data inside the Flash

* **I want** a small portion of the Flash to be writable from the software hosted by the cart,
* **In order to** provide e.g. games on cartridge with advanced functionnality like trophies, save points, and so on.

**Comment by sporniket** [comment]
# Note -- What about non-volatile ram ?

A quick search on mouser, I find 2 solutions : 

* ready made package (all in one : controller, SRAM memory and battery). I don’t like the battery included in package (may be designed as replaceable but not always), plus it’s lithium based battery.
* create a circuit with static ram, NV ram  controller, and one or two backup batteries. With 2 backup batteries, replacement can be performed without cutting backup supply, but anyway, requires big enough batteries, so lithium again

In the end, the more sustainable solution (no need for a backup power supply when the cartridge is not plugged or when the computer is switched off) is to use the Flash memory itself. 

The program MUST implement the writing enable sequence, and the  CPLD MUST deny writing when low pages are selected. E.g. for a 128kB EEPROM (8 pages of 16kb memory), writing should be allowed only on the last two pages.

OR the other way around, allow writing ONLY on the first 2 pages (as a CPLD cannot probe for the size of the EEPROM, and anyway it simplifies the management of writing)

**Comment by sporniket** [comment]
# Candidate design

Leveraging the solution described here : [Programmer une mémoire Flash (EEPROM) 32 broches avec un Arduino](https://elrindel.github.io/programmateur-eeprom-diy-arduino) ; and [stored on github](https://github.com/Elrindel/Arduino-Flash-Programmer)

## Principles

The cartridge would have 3 parts : 
* Data storage and retrieval
* Interface with the host computer (Thomson MO5)
* In system programming

### Data storage and retrieval

This is just the EEPROM (512 kiB - 19 address lines, 8 data lines, 3 control lines).

Power supply comes from the host computer

### Interface with the host computer (Thomson MO5)

> * The host computer has 16KiB addressable through the cartridge port, thus the storage will be seen as 32 banks of 16KiB
>   * A given banks has less than 16KiB of free data though, as the first bytes of a bank MUST contains bank switching code to be executed when a power on reset happens, and the last bytes must specify an entry point
>   * the EEPROM works by blocks of 4KiB, thus a bank will be made of 4 blocks, from 0 to 3
> * Bank switching will be performed by writing a byte to special address
> * Writing commands from the computer is allowed only when the current page is 0 or 1 (one bit) ; or only when the current page is from 0 to 3 (2 bits) ; or only when the current page is from 0 to 7 (3 bits) ; the actual writable capacity for a given page is less than 16KiB though, in effect only bloc 1 and 2 are freely writable, thus 8 KiB per bank with writing allowed
> * When the storage is being reprogrammed by the in system programming part, the storage MUST be disconnected from the computer

All those principles would be addressed with the following design : 

* Enough bidirectionnal buffers to connect/disconnect the storage from the host computer
* A GAL 22v10 will be in charge of bank switching
  * Needs the 16 address lines + write enable (16 inputs) and 5 outputs
* A GAL 22v10 will be in charge of allowing write (5 lines of bank id + write enable + chipselect) and controlling the buffers

### In system programming

* Basically, the equivalent of the Arduino Nano.
* 3 shift registers to address the 19 address lines of the storage and provide 6 supplemental output lines
* One or more status leds
* All connection to the EEPROM and the interface to the host computer is done through voltage level shifters.

> Noteworthy : 
> * Contrary to the inspiring project, the shift registers will be addressed using SDA/SCL lines of the Arduino, and the programming model will be of a 24bits value.
> * The 3 GPIO freed from controlling the address lines will be used to : 
>   * signal to the interface with the host computer the activation/sleep of the programming system
>   * control one or two status LED to indicate erase/write/verify/none
>

**Comment by sporniket** [comment]
# NEW User story

**As** a software developper
**I want** to upload my program on a cartridge that may not be plugged on the host computer
**In order to** be able to perform that operation without being obliged to setup and switch on the computer


# Power supply and isolation of the data storage and retrieval part

* A classical power supply selection using MOSFETs will select the power source of the EEPROM : power from the host computer is priorized before power from VBUS (USB port)
* A set of voltage translators buffers between the host computer and the data storage, another set will buffers between the data storage and the in system programming. The voltage of those buffers on the data storage side will be from the line selected by the power supply selection circuit. 

```
[host computer]------>[interface to host]
|------>[data storage/host buffers]<--/
|           |     |
|           |     \---------[data storage]
|           |     /-------->[            ]
|           |     |              |
\---->[power selection]          |
/---->[               ]          |
|           |                    |
|-------->[data storage/ISP]-----/
|               |
[VBUS]--[ISP]---/
```
